### PR TITLE
Add mean.dtype_out

### PR DESCRIPTION
**Pull request by larryliu0820**
Give it a try and see if CI is happy

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -3916,11 +3916,10 @@
...
 # For normal naming convention this should be `mean.out`. However since we already have `mean.out` we have to rename this.
-# FIXME: fix CI jobs and re-enable this
-#- func: mean.dtype_out(Tensor self, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
-#  device_check: NoCheck   # TensorIterator
-#  dispatch:
-#    CompositeExplicitAutograd: mean_dtype_out
+- func: mean.dtype_out(Tensor self, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
+  device_check: NoCheck   # TensorIterator
+  dispatch:
+    CompositeExplicitAutograd: mean_dtype_out
```
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:3922]: you also need to manually author a kernel in  the `at::native::` namespace, with this name (`mean_dtype_out`). Probably in this file: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/ReduceOps.cpp#L1401

Comment (larryliu0820) [reply aten/src/ATen/native/native_functions.yaml:3922]: Yep added

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -1411,6 +1411,10 @@ Tensor& mean_out(const Tensor& self, DimnameList dim,
   return at::mean_out(result, self, dimnames_to_positions(self, dim), keepdim, opt_dtype);
 }
 
+Tensor& mean_dtype_out(const Tensor &self, std::optional<ScalarType> dtype, Tensor& result) {
+  return at::mean_out(result, self, IntArrayRef{}, false, dtype);
```
Comment (bdhirsh) [inline aten/src/ATen/native/ReduceOps.cpp:1415]: Looking at the CI failure here: https://github.com/pytorch/pytorch/actions/runs/10410262594/job/28833890475?pr=133506

I tried running a similar OpInfo test that passes (`python test/test_ops.py TestCommonCUDA.test_out_stack_cuda_float32`) and seeing how that exception is properly being raised.

For `stack()`, it looks like it comes from this check in the meta function: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorShape.cpp#L284

So you can try adding a similar assert in your kernel here?

**Review by jackzhxng — COMMENTED** [review]
This seems close to landing, will be closing https://github.com/pytorch/pytorch/pull/133533

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -1411,6 +1411,15 @@ Tensor& mean_out(const Tensor& self, DimnameList dim,
   return at::mean_out(result, self, dimnames_to_positions(self, dim), keepdim, opt_dtype);
 }
 
+Tensor& mean_dtype_out(const Tensor &self, std::optional<ScalarType> dtype, Tensor& result) {
+  TORCH_CHECK(
+    canCast(self.scalar_type(), result.scalar_type()),
```
Comment (bdhirsh) [inline aten/src/ATen/native/ReduceOps.cpp:1416]: oh, the test is failing for you because:

(1) `torch.mean(..., out=)` can desugar to 1 of 2 overloads (the old one, and the new one you added)

(2) the OpInfo tests use `torch.mean`, so they run on **both** out= overloads

(3) you've added the error for this new overload, but the old mean doesn't have the error

Comment (bdhirsh) [reply aten/src/ATen/native/ReduceOps.cpp:1416]: The error seems good to have. It would also be nice to add the error to the other overload (5 lines above this one), so we can avoid skipping the test. Although there is a tiny BC risk (we would probably just want to label the module BC breaking, so we can mention it in the release notes).

I think given the tiny BC risk, I'm ok with either option:

(1) add the error to both overloads and fix the test

(2) keep the error just in the new op, but leave the test xskipp'd.

**Review by bdhirsh — APPROVED** [review]

File: aten/src/ATen/native/ReduceOps.cpp (hunk 3 above)
Comment (larryliu0820) [reply aten/src/ATen/native/ReduceOps.cpp:1416]: ok had to go with (2) since (1) doesn't fix the issue. I'm having a hard time to tell if the test is actually hitting the kernels. I added the error message in the skip line.
